File: pkg/sandboxclient/envd.go
package sandboxclient

import (
	"context"
	"errors"
	"net/http"
)

const EnvdPort uint32 = 49983

// EnvdEndpoint is the hand-off contract for an E2B SDK/Connect client. Fast
// Sandbox resolves and authenticates the network route but deliberately does
// not redefine envd's Process or Filesystem protobuf services.
type EnvdEndpoint struct {
	BaseURL string
	Headers http.Header
	Route   Route
}

type EnvdAdapter struct {
	Resolver RouteResolver
	Port     uint32
}

func (a *EnvdAdapter) Resolve(ctx context.Context, sandbox SandboxRef) (EnvdEndpoint, error) {
	if a == nil || a.Resolver == nil {
		return EnvdEndpoint{}, errors.New("Envd adapter route resolver is not configured")
	}
	port := a.Port
	if port == 0 {
		port = EnvdPort
	}
	route, err := a.Resolver.Resolve(ctx, sandbox, port)
	if err != nil {
		return EnvdEndpoint{}, err
	}
	return EnvdEndpoint{BaseURL: route.Endpoint.String(), Headers: route.RequiredHeaders.Clone(), Route: route}, nil
}

File: sdk/python/fast_sandbox/envd.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping

from .route import EndpointResolver

ENVD_PORT = 49983


@dataclass(frozen=True)
class EnvdEndpoint:
    """Route hand-off for the official E2B envd Connect client."""

    base_url: str
    headers: Mapping[str, str]
    sandbox_uid: str
    route_generation: int


class EnvdAdapter:
    def __init__(self, resolver: EndpointResolver, port: int = ENVD_PORT):
        self._resolver = resolver
        self._port = port

    def resolve(self, sandbox_name: str, namespace: str = "") -> EnvdEndpoint:
        route = self._resolver.resolve(sandbox_name, self._port, namespace)
        return EnvdEndpoint(
            base_url=route.endpoint,
            headers=dict(route.headers),
            sandbox_uid=route.sandbox_uid,
            route_generation=route.route_generation,
        )
